src/Parser/GuardianDruid/Modules/Features/AntiFillerSpam.js

```diff
@@ -0,0 +1,199 @@
+
+import SPELLS from 'common/SPELLS';
+import { formatPercentage } from 'common/format';
+import SpellIcon from 'common/SpellIcon';
+import SpellLink from 'common/SpellLink';
+
+import StatisticBox from 'Main/StatisticBox';
+
+import Module from 'Parser/Core/Module';
+import Combatants from 'Parser/Core/Modules/Combatants';
+import EnemyInstances from 'Parser/Core/Modules/EnemyInstances';
+import DynamicHaste from './DynamicHaste';
+
+import ActiveTargets from './ActiveTargets';
+
+const debug = false;
+
+// The  amount of time after a proc has occurred when casting a filler is no longer acceptable
+const REACTION_TIME_THRESHOLD = 500;
+
+class AntiFillerSpam extends Module {
+  static dependencies = {
+    combatants: Combatants,
+    enemyInstances: EnemyInstances,
+    dynamicHaste: DynamicHaste,
+    activeTargets: ActiveTargets,
+  };
+
+  abilityLastCasts = {};
+  _totalGCDSpells = 0;
+  _totalFillerSpells = 0;
+  _unnecessaryFillerSpells = 0;
+
+  on_byPlayer_cast(event) {
+    const spellID = event.ability.guid;
+    const spell = GCD_SPELLS[spellID];
+    if (!spell) {
+      return;
+    }
+
+    this._totalGCDSpells += 1;
+    const timestamp = event.timestamp;
+    const spellLastCast = this.abilityLastCasts[spellID] || -Infinity;
+    const targets = this.activeTargets.getActiveTargets(event.timestamp).map(enemyID => this.enemyInstances.enemies[enemyID]).filter(enemy => !!enemy);
+    const combatant = this.combatants.selected;
+    this.abilityLastCasts[spellID] = timestamp;
+
+    const isFillerEval = (typeof spell.isFiller === 'function') ? spell.isFiller(event, combatant, targets, spellLastCast) : spell.isFiller;
+    if (!isFillerEval) {
+      return;
+    }
+
+    debug && console.group(`[FILLER SPELL] - ${spellID} ${SPELLS[spellID].name}`);
+
+    this._totalFillerSpells += 1;
+
+    const availableSpells = [];
+    Object.keys(GCD_SPELLS).forEach((gcdSpellID) => {
+      // We have to cast to string here because Object keys are always strings
+      if (`${spellID}` === gcdSpellID) {
+        return;
+      }
+
+      const { isFiller, baseCD, hastedCD, condition, proc, isUtility } = GCD_SPELLS[gcdSpellID];
+      const lastCast = this.abilityLastCasts[gcdSpellID] || -Infinity;
+      const isFillerEval = (typeof isFiller === 'function') ? isFiller(event, combatant, targets, lastCast) : isFiller;
+      if (isFillerEval || isUtility) {
+        return;
+      }
+
+      let isOffCooldown = true;
+      if (baseCD !== null) {
+        if (hastedCD) {
+          isOffCooldown = this.dynamicHaste.isHastedAbilityOffCooldown(lastCast, timestamp, baseCD);
+        } else {
+          isOffCooldown = timestamp - lastCast > baseCD;
+        }
+      }
+
+      const hasProc = proc && ((typeof proc === 'function') ? proc(event, combatant, targets, lastCast) : proc);
+      const meetsCondition = (condition !== undefined) ? ((typeof condition === 'function') ? condition(event, combatant, targets, lastCast) : condition) : true;
```

Comment:
I think it would be much easier to follow if this was written this with short if statements and return early when a required one fails.